### PR TITLE
 resolver: Implement pipe with capture and replace

### PR DESCRIPTION
**Pull request by qinqon**
For the main scenario we have an expression like [1] this expression
uses a capture reference path as pipe in and a replace expression as
pipe out. This change convert pipe as the first argument on the ternary
replace operation.
  
[1] capture.base-iface-routes | routes.running.next-hop-interface:="br1"

### CODE REVIEW

File: tests/basic_test.go [hunk 1]
```diff
@@ -237,7 +228,66 @@ interfaces:
...
+    table-id: 254
+  - destination: 1.1.1.0/24
+    next-hop-address: 192.168.100.1
+    next-hop-interface: br1
+    table-id: 254
+`),
+}
+
+func testPolicyWithFilterCaptureAndDesiredStateCaptureEntryRef(t *testing.T) {
+	t.Run("with a eqfilter capture expression and desired state that has no ref", func(t *testing.T) {
```
Comment (AlonaKaplan) [inline tests/basic_test.go:264]: This sentence seems to be outdated.

File: tests/basic_test.go [hunk 2]
```diff
@@ -258,28 +308,15 @@ func testPolicyWithDesiredStateRefToCachedCapturedStates(t *testing.T) {
 	t.Run("with desired state refereing a cached captured state", func(t *testing.T) {
```
Comment (AlonaKaplan) [inline tests/basic_test.go:308]: `testPolicyWithPartialCache`

File: tests/basic_test.go [hunk 3]
```diff
@@ -329,23 +352,17 @@ func testPolicyWithCaptureEntryExpressionRefToCachedCapturedStates(t *testing.T)
 	t.Run("with a eqfilter capture expression and desired state that has no ref", func(t *testing.T) {
```
Comment (AlonaKaplan) [inline tests/basic_test.go:352]: `testPolicyWithPartialCache2`

File: nmpolicy/internal/parser/parser.go [hunk 4]
```diff
@@ -245,3 +264,14 @@ func (p *parser) parseReplace() error {
 	p.lastNode = operator
 	return nil
 }
+
+func (p *parser) parsePipe() error {
+	if p.lastNode == nil {
+		return invalidPipeError("missing pipe in expression")
+	}
+	if p.lastNode.Path == nil {
```
[the remark this answers is not in the document]
Comment (qinqon) [reply nmpolicy/internal/parser/parser.go:272]: Right now we only need piping at path expressions, let's not open the pandora box yet.

File: nmpolicy/internal/parser/parser.go [hunk 5]
```diff
@@ -228,7 +240,14 @@ func (p *parser) parseReplace() error {
 	if p.lastNode.Path == nil {
 		return invalidReplaceError("left hand argument is not a path")
 	}
-	operator.Replace[0].Terminal = ast.CurrentStateIdentity()
+
+	if p.pipedInNode != nil {
```
[the remark this answers is not in the document]
Comment (qinqon) [reply nmpolicy/internal/parser/parser.go:244]: They are near identical, I will try to has a common function for Replace and EQ filter.

File: tests/basic_test.go [hunk 6]
```diff
@@ -237,7 +228,66 @@ interfaces:
...
+    next-hop-interface: br1
+    table-id: 254
+  - destination: 1.1.1.0/24
+    next-hop-address: 192.168.100.1
+    next-hop-interface: br1
+    table-id: 254
+`),
+}
+
+func testPolicyWithFilterCaptureAndDesiredStateCaptureEntryRef(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (qinqon) [reply tests/basic_test.go:263]: This one has no cache
```
 obtained, err := nmpolicy.GenerateState(                                
            policySpec,                                                         
            mainCurrentState,                                                   
            types.CachedState{})               
```

File: nmpolicy/internal/parser/parser.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (qinqon) [reply nmpolicy/internal/parser/parser.go]: Just added a common for the piped vs currentState thing, we just have a pair of ternary operators, if something else pop up with can start to work on more advanced stuff like parser function table or visitor pattern or the like, let's not add more complexity here.

File: tests/basic_test.go [hunk 7]
```diff
@@ -237,7 +228,66 @@ interfaces:
...
+    table-id: 254
+  - destination: 1.1.1.0/24
+    next-hop-address: 192.168.100.1
+    next-hop-interface: br1
+    table-id: 254
+`),
+}
+
+func testPolicyWithoutCache(t *testing.T) {
+	t.Run("with a replace, eqfilter capture expression and desired state that has ref", func(t *testing.T) {
```
Comment (AlonaKaplan) [inline tests/basic_test.go:264]: I would change the message to the name of the method.

File: tests/basic_test.go [hunk 8]
```diff
@@ -325,27 +348,21 @@ interfaces:
 	})
 }
 
-func testPolicyWithCaptureEntryExpressionRefToCachedCapturedStates(t *testing.T) {
+func testPolicyWithPartialCache(t *testing.T) {
 	t.Run("with a eqfilter capture expression and desired state that has no ref", func(t *testing.T) {
```
Comment (AlonaKaplan) [inline tests/basic_test.go:352]: same. The current message is unclear.

File: tests/basic_test.go [hunk 9]
```diff
@@ -254,32 +304,19 @@ interfaces:
 	})
 }
 
-func testPolicyWithDesiredStateRefToCachedCapturedStates(t *testing.T) {
+func testPolicyWithFullCache(t *testing.T) {
 	t.Run("with desired state refereing a cached captured state", func(t *testing.T) {
```
Comment (AlonaKaplan) [inline tests/basic_test.go:308]: same

**Review by AlonaKaplan — APPROVED** [review]

**Pull request merged**